[UI] Add timespan selection and auto refresh for metrics tab in node page

ChengYanJin opened the pull request:
**Component**: ui, nodes

**Context**:  Add timespan selection and auto-refresh for metrics tab in the Node page

**Summary**:
- Add timespan selection in `metrics` tab, by default is Last 24 hours
- Add selected timespan in URL query
- Add auto-fresh of graphs every 1 min
- Keep the selected timespan when we switch back to metrics tab
- When we select a node from the list, we don't add the default the timespan(from=now-24h) in the URL. 
- The style of dropdown component is not the same as design. (Need to work on it in Core-ui)
- Launch a long-time running saga at the root of saga to watch the Refresh_NODESTATS action.

**Acceptance criteria**: 
![image](https://user-images.githubusercontent.com/18453133/93862114-e0e1bf00-fcc1-11ea-9d6d-49cf13e0f08f.png)

![image](https://user-images.githubusercontent.com/18453133/93861686-36699c00-fcc1-11ea-91b3-2c8358ad2a34.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2776 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/ducks/monitoring: Add fetch node stats for the metrics tab

Refs: #2776

ChengYanJin committed:
ui/nodes/metrics: Add timespan selection for the graphs

Refs: #2776

ChengYanJin committed:
ui/nodes: Add the selected timespan in the URL

- When switch bwtween the nodes, keep the same tab, and same timespan if metric tab is selected
- When switch bwtween the nodes, if there no tab selected, by default `health` tab is selected

Refs: #2776

ChengYanJin committed:
ui/nodes: Add the auto-refresh for metrics graphs

Refs: #2776

ChengYanJin committed:
ui/nodes: When switch back to metrics tab in Node Page, keep the same timespan selected

Refs: #2776

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

ChengYanJin committed:
ui/nodes: Set the height and width of the graph according to the current size of screen

Refs: #2776

ChengYanJin committed:
ui/package: Update the core-ui version

Add a new prop activeColor for Tabs component

Refs: #2776

ChengYanJin committed:
ui/nodes: Customize the x-axis label according to the selected timespan

Refs: #2776

ChengYanJin committed:
ui/nodes: Move NodePageMetricsTab from dumb component to container

Refs: #2776

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page`
* Rebase `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` into
`development/2.6`
* [w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

bert-e commented:
# History mismatch



Merge commit #a4d221a09f14dd7eecbd8680d9b72b430013004b on the integration branch
`w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` is merging a branch which is neither the current
branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` into
`development/2.6`
* [w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -212,8 +239,48 @@ const NodePageMetricsTab = (props) => {
+      label: LAST_SEVEN_DAYS,
+      onClick: () => {
+        dispatch(updateNodeStatsAction({ metricsTimeSpan: LAST_SEVEN_DAYS }));
+        updateMetricsGraph();
```
Really more a question than a suggestion, but wouldn't it be "cleaner" to handle this at the saga level?

Instead of a single "refresh" saga which is going to send this "update metrics" action periodically, we could have it use a `race` effect, between `take(TIMESPAN_UPDATED)` and `delay(REFRESH_PERIOD)`, making sure we won't have too many call sites for the chart update... WDYT?

ChengYanJin replied:
Definitely right! 
Will dispatch `updateNodeStatsFetchArgumentAction` to the `race` everytime we change the timespan or node(information)

gdemonet commented on ui/src/components/NodeListTable.js:
```diff
@@ -390,8 +412,12 @@ const NodeListTable = (props) => {
+    if (isTabSelected && location.pathname.endsWith('metrics')) {
+      // Keep the same timespan selected for metrics tab, when switch between the nodes
+      const currentTab = location?.pathname?.split('/')?.pop();
+      history.push(`/newNodes/${nodeName}/${currentTab}?from=${queryTimeSpan}`);
```
This means we loose the other query parameters (like the table filters)... But I see it's not just related to this change, because we had this bug before apparently.

Wouldn't it be simpler to just keep the exact same URL and switch only the Node name part?
Something like:

```javascript
const newPath = location.pathname.replace(/\/newNodes\/[^\/]*\//, `/newNodes/${nodeName}/`);
history.push(newPath);
```

If you really need to sanitize the query string, I think you'd be able to by using `URLSearchParams`.

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -245,6 +251,9 @@ const NodePageMetricsTab = (props) => {
         updateMetricsGraph();
+        history.push(
+          `/newNodes/${selectedNodeName}/metrics?from=${QUERY_LAST_SEVEN_DAYS}`,
+        );
```
Ditto, this approach cleans off the other parts of the querystring, which is not desired behaviour IMO

gdemonet commented on ui/src/containers/NodePageRSP.js:
```diff
@@ -53,7 +56,7 @@ const NodePageRSP = (props) => {
   const query = useQuery();
   const queryTimespan = query.get('from');
-  let metricsTimeSpan;
+  let metricsTimeSpan = LAST_TWENTY_FOUR_HOURS;
```
Why this change? The `default` case below seems sufficient

ChengYanJin replied:
Actually, and we set the default value in the saga state. 
Maybe it doesn't hurt to set again here. 

``` javascript
  const query = useQuery();
  const nodeMetricsTimeSpan = useSelector(
    (state) => state.app.monitoring.nodeStats.metricsTimeSpan,
  );

  let metricsTimeSpan;
  const queryTimespan = query.get('from');

  if (queryTimespan) {
    // If timespan query specified in query string
    metricsTimeSpan = queryTimeSpansCodes?.find(
      (timespan) => timespan.label === queryTimespan,
    )?.value;
  } else {
    metricsTimeSpan = nodeMetricsTimeSpan;
  }
```


```javascript
nodeStats: {
    metricsTimeSpan: LAST_TWENTY_FOUR_HOURS,
    ...
}
```

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -87,6 +89,9 @@ const defaultState = {
   },
   nodeStats: {
     metricsTimeSpan: LAST_TWENTY_FOUR_HOURS,
+    instanceIP: '',
+    controlPlaneInterface: '',
+    workloadPlaneInterface: '',
```
I think it makes sense to maintain these values in the state (in essence, this represents the "selected node" for which we'll pull metrics). However, doing this should allow us to only set them once, when we mount the metrics tab component (because we know which Node is selected), and never touch them again.

In practice, it means that:

- when mounting / rendering the component, you dispatch this once:

```javascript
      updateNodeStatsAction({
        metricsTimeSpan,
        instanceIP,
        controlPlaneInterface,
        workloadPlaneInterface,
      })
```

- the fetch saga reads from the state this information about the selected node and timespan, not from the payload, so you dispatch this at render, after the previous one:

```javascript
    fetchNodeStatsAction()
```

- the refresh saga doesn't care about the state, it's only periodically sending a `FETCH_NODE_STATS` action until it receives a halt (see later comment on how to write the refresh saga)

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -725,6 +743,33 @@ export function* fetchNodeStats({ payload }) {
+  if (isRefreshing) {
+    yield call(refreshNodeStats);
+  }
+}
```
This pattern for refresh sagas was already discussed with Carlito, back in the days, and we came to the conclusion that it had big drawbacks (you can loose a "STOP" action, or trigger the refresh loop multiple times in parallel, and overall, it's not easy to read). See #2035.

In practice, with the suggestion from a previous comment, that would look like:

```javascript
export function* refreshNodeStats() {
  while (true) {
    yield take(REFRESH_NODESTATS);
    while (true) {
      yield fork(fetchNodeStats);
      const {interrupt, update} = yield race({
        interrupt: take(STOP_REFRESH_NODESTATS),
        // If the refresh period expires before we receive a halt,
        // we can refresh the stats
        requeue: delay(DELAY_TIMEOUT),
        // NOTE: This action doesn't exist, but it should, so that
        // whenever we change one of the parameters for "fetchNodeStats",
        // it gets triggered again
        update: take(UPDATE_SELECTED_NODE),
      });
      if (interrupt) {
        yield cancel(fetchNodeStats);
        break;
      }
    }
  }
}
```

ChengYanJin replied:
hello @gdemonet ,

I wanted to address this dept ticket for a long time! Let's work on it now!

I have one questions in my mind:
What does `yield take(REFRESH_NODESTATS);` do? In this case, I think we may never have the `REFRESH_NODESTATS` dispatched.

gdemonet replied:
`take(SOME_EVENT)` is blocking until `SOME_EVENT` is seen.

So, the idea is to replace the `takeFirst` effect creator:

```javascript
while (true) {
  // wait for the start event
  yield take("START_REFRESH");
  // once we've seen this start event, we can enter the refresh loop
  while (true) {
    // trigger the data update
    yield fork(fetchData);
    // wait for the stop event for at most the refresh period
    const {interrupt} = yield race({
      interrupt: take("STOP_REFRESH"),
      requeue: delay(REFRESH_PERIOD),
    });
    if (interrupt) {
      // we received the stop event, so we get out of the refresh loop
      // and we will wait for the next start event
      break;
    }
  }
}
```

gdemonet commented on ui/src/containers/NodePageRSP.js:
```diff
@@ -53,19 +54,38 @@ const NodePageRSP = (props) => {
+      default:
+        queryTimespan = QUERY_LAST_TWENTY_FOUR_HOURS;
+    }
   }
```
Would be nicer to use maps, especially if we start adding more options.

gdemonet reviewed (changes requested):
Just some code review (mostly Sagas again :wink:), I'll test it soon!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui/nodes: When select another node replace only with the node name

ChengYanJin committed:
ui/nodes: Fix the bug in metrics tab clean off other parts of the querystring

ChengYanJin committed:
ui/ducks/monitoring: Remove the payload for fetchNodeStatsAction

When rendening the component, we dispatch the updateNodeStatsAction once with the payload.
Then we keep the information about the selected node in the state.
So when fetchNodeStatsAction, there is no need to carry the payload again.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page`).

```bash
 $ git fetch
 $ git checkout origin/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page
```

ChengYanJin committed:
ui/ducks/monitoring: Improve the refresh saga

ChengYanJin committed:
ui/ducks/monitoring: Add the UPDATE_NODESTATS_FETCH_ARG action

ChengYanJin committed:
ui/nodes: Extract the queryTimeSpansCodes to constant

bert-e commented:
# History mismatch



Merge commit #652a67202917b08a128b51b7fb937758574a1072 on the integration branch
`w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` is merging a branch which is neither the current
branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin committed:
ui/podes: Fix the bug

gdemonet commented on ui/src/components/NodeListTable.js:
```diff
@@ -391,12 +390,14 @@ const NodeListTable = (props) => {
-      const currentTab = location?.pathname?.split('/')?.pop();
-      history.push(`/newNodes/${nodeName}/${currentTab}`);
+      const newPath = location.pathname.replace(
+        // eslint-disable-next-line no-useless-escape
```
If there is a useless escape, maybe we should fix it?

ChengYanJin replied:
It's actually an [eslint-error](https://stackoverflow.com/a/47277197) because of the regex: Unnecessary escape character: \/ no-useless-escape 
But I think the regex is correct. That's why I disable it, maybe the message is not easy to understand.

slaperche-scality replied:
As mentioned in the answer you linked, you don't need to escape the forward slash inside the character class e.g. you could write `/\/newNodes\/[^/]*\//` and have the same behavior:

```
> var nodeName = "foo";
undefined
> var str = "/newNodes/BAR/";
undefined
> str.replace(/\/newNodes\/[^\/]*\//, `/newNodes/${nodeName}/`); 
'/newNodes/foo/'
> str.replace(/\/newNodes\/[^/]*\//, `/newNodes/${nodeName}/`); 
'/newNodes/foo/'
```

Did you try that version?
Is ESLint happy with this regex? Or do you still have the warning?

ChengYanJin replied:
I confirm `/\/newNodes\/[^/]*\//` is working correctly!

gdemonet commented on ui/src/components/NodeListTable.js:
```diff
@@ -391,12 +390,14 @@ const NodeListTable = (props) => {
     } else {
-      // Set Health tab as default tab
-      history.push(`/newNodes/${nodeName}/health`);
+      history.push({ pathname: `/newNodes/${nodeName}/health` });
```
We'll loose URL search params here as well :wink: 
In any case, I think this URL parsing in the component is not the right approach, we should instead rely on react-router's `Route` matching, that'd be much cleaner and safer. Let's open a debt ticket for this, we'll try to refactor other pages at the same time.

gdemonet replied:
#2838

ChengYanJin replied:
You are right! Thanks for creating the debt ticket!
We're gonna lose the search params if we want to add filters for the table.

gdemonet commented on ui/src/containers/NodePageMetricsTab.js:
```diff
@@ -57,11 +68,22 @@ const LoaderContainer = styled(Loader)`
+  const dispatch = useDispatch();
   const theme = useSelector((state) => state.config.theme);
+  const history = useHistory();
+  const query = new URLSearchParams(history?.location?.search);
```
Is there a difference between this and the `useQuery` hook? Why do we need to rely on `useHistory().location` rather than `useLocation()` directly?

gdemonet commented on ui/src/containers/NodePageMetricsTab.js:
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
+      onClick: () => {
+        dispatch(
+          updateNodeStatsAction({ metricsTimeSpan: LAST_SEVEN_DAYS }),
+          updateNodeStatsFetchArgumentAction(),
```
Would be a bit cleaner to differenciate between the two:

- `UPDATE_NODESTATS` is used to update `app.monitoring.nodeStats.metrics`
- `UPDATE_NODESTATS_FETCH_ARG` is used to update the other keys used as fetch arguments

This way, we could:

- only dispatch one action, instead of two
- never forget to send the `UPDATE_FETCH_ARG` action when needed, because that's the only way to actually update them
- rely on the main `UPDATE_NODESTATS` if we had other sagas that need to react on some metrics update

gdemonet commented on ui/src/containers/NodePageMetricsTab.js:
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
+      },
+      selected: metricsTimeSpan === LAST_ONE_HOUR,
+    },
+  ];
```
Could be refactored:

```suggestion
  const metricsTimeSpanItems = [
    LAST_SEVEN_DAYS, LAST_TWENTY_FOUR_HOURS, LAST_ONE_HOUR
  ].map(
    option => ({
      label: option,
      onClick: () => {
        dispatch(
          updateNodeStatsFetchArgumentAction({ metricsTimeSpan: option }),
        );
        writeUrlTimeSpan(option);
      },
      selected: metricsTimeSpan === option,
    })
  );
```

gdemonet commented on ui/src/containers/NodePageMetricsTab.js:
```diff
@@ -212,8 +236,65 @@ const NodePageMetricsTab = (props) => {
+    },
+  ];
+
+  const metricsTimeSpanDropdownItems = metricsTimeSpanItems?.filter(
+    (mTS) => mTS.label !== metricsTimeSpan,
+  );
```
I'm surprised, does it mean we only have one item in the dropdown?

ChengYanJin replied:
Nope. The items we have in the dropdown are all the options **exclude** the current selected`metricsTimeSpan`

gdemonet replied:
:man_facepalming: Oopsie, I read that too fast! Thanks for the translation :)

gdemonet reviewed (changes requested):
Mostly a change on how we handle updating the "fetch args"

ChengYanJin committed:
ui/ducks/monitoring: Dispatch UPDATE_NODESTATS action only to update the metrics in nodeStats

UPDATE_NODESTATS is used to update `app.monitoring.nodeStats.metrics`
UPDATE_NODESTATS_FETCH_ARG is used to update the other keys used as fetch arguments

ChengYanJin committed:
ui/nodes: Use UseQuery hook to parse the query string

ChengYanJin committed:
ui/nodes: Fix the regex and history.push with the query

gdemonet reviewed (approved):
LGTM!

ChengYanJin commented:
/approve

bert-e commented:
# History mismatch



Merge commit #652a67202917b08a128b51b7fb937758574a1072 on the integration branch
`w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` is merging a branch which is neither the current
branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page` into
`development/2.6`
* [w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/05777a78443ef0ec2d9e48562adcb6cd795ed3e0) did not succeed in branch w/2.7/improvement/add-timespan-selection-and-auto-refresh-for-metrics-tab-in-node-page.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye chengyanjin.
